### PR TITLE
chore(26181): OpenAPI specs for the bi-directionality features

### PR DESCRIPTION
**Pull request by vanch3d**
See https://hivemq.kanbanize.com/ctrl_board/57/cards/26181/details/

This PR is part of the bi-directional epic and adds a proposal for the expected OpenAPI specs of the new features

The changes are as follows:

- `WRITE` added to the capability of the protocol adapters
- Added CRUD operations for handling `tags` on devices (through their protocol adapter)
- Added a new `Client` category of services to handle requests for MQTT client filters
- Added a new `Domain` category of services to handle requests for domain tags and topic schema

For the definition and scope of the entities, refer to the `bi-directionality` MIRO board: https://miro.com/app/board/uXjVKydZ7vo=/#

### Problem Details
The PR also corrects the definition of the `ProblemDetails` specification, [RFC9457](https://www.rfc-editor.org/info/rfc9457) (formerly [RFC7807](https://tools.ietf.org/html/rfc7807)).

The `OpenAPI` specs are copied from the official SmartBear registry, https://github.com/SmartBear-DevRel/problems-registry. They are embedded in the specs but, ideally, they should be used as a reference to an external file. The specs have been added as a separate `OpenAPI` specs for reference.

### Out-of-scope
- The specification of the bi-directional mapping cannot be represented, as it is "embedded" in the adapters' undocumented `config` property. A custom endpoint could be created to export the mappings; it will be tried in a further ticket.
- HiveMQ should define its own Problem Details registry to maximise reuse across products. It will be described in another ticket. 

### Design

For clarity, a version of the specs containing only the changes can be found below, to be used in Swagger or IntelliJ.

```yaml
openapi: 3.0.1
info:
  title: HiveMQ Edge REST API
  version: "2024.8-proposal"
tags:
  - description: Interact with protocol adapters.
    name: Protocol Adapters
  - name: Domain
    description: Interact with the elements of the domain conceptualised around the Edge instance
  - name: Client
    description: Interact with MQTT clients connected to the Edge Broker

paths:
  /api/v1/management/protocol-adapters/adapters/{adapterId}/tags:
    get:
      description: Get the domain tags for the device connected through this adapter
      operationId: get-adapter-domainTags
      parameters:
        - description: The name of the adapter to query.
          in: path
          name: adapterId
          required: true
          schema:
            type: string
      responses:
        "200":
          description: Success
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/DomainTagList'
              examples:
                OPC-UA:
                  # Example OPC-UA data point
                  value:
                    - tag: "opc-ua/pump1/temperature"
                      dataPoint:
                        nodeId: "ns=3;i=1002"
                MODBUS:
                  # Example MODBUS data point
                  value:
                    - tag: "modbus/pump2/temperature"
                      dataPoint:
                        start: 1
                        shift: 24
      summary: Get the domain tags for the device connected through this adapter
      tags:
        - Protocol Adapters
    post:
      description: Add a new domain tag to the specified adapter
      operationId: add-adapter-domainTags
      parameters:
        - description: The adapter type.
          in: path
          name: adapterId
          required: true
          schema:
            type: string
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/DomainTag'
        description: The new domain tag.
        required: true
      responses:
        "200":
          description: Success
      summary: Add a new  domain tag
      tags:
        - Protocol Adapters
  /api/v1/management/protocol-adapters/adapters/{adapterId}/tags/{tagId}:
    delete:
      description: Delete the specified domain tag on the given adapter.
      operationId: delete-adapter-domainTags
      parameters:
        - description: The adapter Id.
          in: path
          name: adapterId
          required: true
          schema:
            type: string
        - description: The tag id.
          in: path
          name: tagId
          required: true
          schema:
            type: string
      responses:
        "200":
          description: Success
      summary: Delete a domain tag
      tags:
        - Protocol Adapters
    put:
      description: Update the domain tag of an adapter
      operationId: update-adapter-domainTags
      parameters:
        - description: The adapter Id.
          in: path
          name: adapterId
          required: true
          schema:
            type: string
        - description: The tag id.
          in: path
          name: tagId
          required: true
          schema:
            type: string
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/DomainTag'
      responses:
        "200":
          description: Success
      summary: Update a domain tag
      tags:
        - Protocol Adapters
 
  /api/v1/management/domain/tags:
    get:
      description: Get the list of all domain tags created in this Edge instance
      operationId: get-domain-tags
      responses:
        "200":
          content:
            application/json:
              schema:
                type: array
                items:
                  type: string
          description: Success
      summary: Get all domain tags
      tags:
        - Domain
  /api/v1/management/domain/tags/schema:
    get:
      description: Get the data schema associated with the specified tags
      operationId: get-tag-schemas
      parameters:
        - description: The list of tags to query the schema from
          in: query
          name: tags
          required: true
          schema:
            type: array
            items:
              type: string
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TagSchema'
          description: Success
      summary: Get data schemas
      tags:
        - Domain
  /api/v1/management/domain/topics:
    get:
      description: Get the list of all topic tags created in this Edge instance
      operationId: get-domain-topics
      responses:
        "200":
          content:
            application/json:
              schema:
                type: array
                items:
                  type: string
          description: Success
      summary: Get all topic tags
      tags:
        - Domain
  /api/v1/management/domain/topics/schema:
    get:
      description: Get the data schema associated with the specified topics
      operationId: get-topic-schemas
      parameters:
        - description: The list of topics to query the schema from
          in: query
          name: topics
          required: true
          schema:
            type: array
            items:
              type: string
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/TagSchema'
          description: Success
      summary: Get data schemas
      tags:
        - Domain

  /api/v1/management/client/topic-samples:
    get:
      description: Get a sample of MQTT topics published by clients connected to the Edge broker over a period of time
      operationId: get-client-topics
      parameters:
        - description: The time limit for the observation.
          in: query
          name: queryTime
          required: true
          schema:
            type: integer
            minimum: 500
            default: 1000
            maximum: 10000
      responses:
        "200":
          content:
            application/json:
              example:
                items:
                  - client1/topic1
                  - client1/topic2
                  - client2/topic1
              schema:
                $ref: '#/components/schemas/ClientTopicList'
          description: Success
      summary: Get a sample of MQTT topics
      tags:
        - Client
  /api/v1/management/client/filters:
    get:
      description: Get a list of all clients filters created for the Edge instance
      operationId: get-client-filters
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ClientFilterList'
          description: Success
      summary: Get a list of MQTT client filters
      tags:
        - Client
    post:
      description: Add a client filter to the Edge instance
      operationId: add-client-filter
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/ClientFilter'
        description: The new client filter
        required: true
      responses:
        "200":
          description: Success
      summary: Add a client filter
      tags:
        - Client
  /api/v1/management/client/filters/{clientFilterId}:
    get:
      description: Get the details of the specified client filter and its topcis
      operationId: get-client-filter
      parameters:
        - description: The client filter Id.
          in: path
          name: clientFilterId
          required: true
          schema:
            type: string
      responses:
        "200":
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/ClientFilter'
          description: Success
      summary: Get a MQTT client filter
      tags:
        - Client
    delete:
      description: Delete the specified client filter.
      operationId: delete-client-filter
      parameters:
        - description: The client filter Id.
          in: path
          name: clientFilterId
          required: true
          schema:
            type: string
      responses:
        "200":
          description: Success
      summary: Delete a client filter
      tags:
        - Client
    put:
      description: Update the specified client filter
      operationId: update-client-filter
      parameters:
        - description: The client filter Id.
          in: path
          name: clientFilterId
          required: true
          schema:
            type: string
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/ClientFilter'
      responses:
        "200":
          description: Success
      summary: Update a client filter
      tags:
        - Client
components:
  schemas:
    QoS:
      type: integer
      format: int32
      default: 0
      description: The maxQoS for this subscription.
      enum:
        - 0
        - 1
        - 2
    ClientFilterConfiguration:
      type: object
      required:
        - destination
      properties:
        destination:
          type: string
        maxQoS:
          $ref: '#/components/schemas/QoS'
    ClientFilter:
      type: object
      description: A client filter
      properties:
        id:
          type: string
          description: The unique id of the client filter
        topicFilters:
          type: array
          description: The list of topics associated with this client filter
          items:
            $ref: '#/components/schemas/ClientFilterConfiguration'
      required:
        - id
        - topicFilters
    ClientFilterList:
      type: array
      description: List of result items that are returned by this endpoint
      items:
        $ref: '#/components/schemas/ClientFilter'
    DeviceDataPoint:
      ### TODO We need to name that entity
      type: object
      description: A data point on a device connected to the adapter
    DomainTag:
      type: object
      description: A tag associated with a data point on a device connected to the adapter
      properties:
        tag:
          type: string
          description: The Tag associated with the data-point.
        dataPoint:
          $ref: '#/components/schemas/DeviceDataPoint'
      required:
        - tag
        - dataPoint
    DomainTagList:
      type: object
      properties:
        items:
          type: array
          description: List of result items that are returned by this endpoint
          items:
            $ref: '#/components/schemas/DomainTag'
    ClientTopicList:
      type: object
      properties:
        items:
          type: array
          description: List of result topics that are returned by this endpoint
          items:
            type: string
  
    ProtocolAdapter:
      type: object
      description: List of result items that are returned by this endpoint
      properties:
        author:
          type: string
          description: The author of the adapter
        capabilities:
          type: array
          description: The capabilities of this adapter
          items:
            type: string
            description: The capabilities of this adapter
            enum:
              - WRITE
              - READ
              - DISCOVER
          uniqueItems: true
    TagSchema:
      type: object
      description: The JSON-Schema specification of a data point
```

### CODE REVIEW

**Comment by sfrehse** [comment]
@vanch3d, it's great that you come up with the proposal. Do you have any idea how we can better review that? The view in GH is not optimal to comment :-)

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 1]
```diff
@@ -0,0 +1,5561 @@
...
+    get:
+      description: Get the data schema associated with the specified tags
+      operationId: get-tag-schemas
+      parameters:
+        - description: The list of tags to query the schema from
+          in: query
+          name: tags
+          required: true
+          schema:
+            type: array
```
Comment (DC2-DanielKrueger) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:3948]: Is this really an array of tags? The returned schema is just a single one, so all endpoints need to have the same schema?

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 2]
```diff
@@ -0,0 +1,5561 @@
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TagSchema'
+          description: Success
+      summary: Get data schemas
+      tags:
+        - Domain
+
+  /api/v1/management/client/topic-samples:
```
Comment (DC2-DanielKrueger) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:4001]: Why is this needed?

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml (hunk 1 above)
Comment (vanch3d) [reply ext/hivemq-edge-openapi-2024.8-proposal.yaml:3948]: Good find. The return value should match the params: if it's a list of tags, then it's a list of schemas, indexed by their tags; if it's a single schema, then the param must be a single tag
Will check most common usage and will update

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 3]
```diff
@@ -0,0 +1,5561 @@
...
+          description: The type of the validator.
+          enum:
+            - SCHEMA
+      required:
+        - arguments
+        - type
+    DeviceDataPoint:
+      ### TODO We need to name that entity
+      type: object
+      description: A data point on a device connected to the adapter
```
Comment (DC2-DanielKrueger) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:4620]: properties missing

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 4]
```diff
@@ -0,0 +1,5561 @@
...
+    DomainTag:
+      type: object
+      description: A tag associated with a data point on a device connected to the adapter
+      properties:
+        tag:
+          type: string
+          description: The Tag associated with the data-point.
+        dataPoint:
```
[the remark this answers is not in the document]
Comment (vanch3d) [reply ext/hivemq-edge-openapi-2024.8-proposal.yaml:4628]: `dataPoint` definitely an object. It's the `address` of the tag on the device
I've added `additionalProperties: true` to the spec

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml (hunk 3 above)
Comment (vanch3d) [reply ext/hivemq-edge-openapi-2024.8-proposal.yaml:4620]: I've added additionalProperties: true to the spec

Comment (vanch3d) [reply ext/hivemq-edge-openapi-2024.8-proposal.yaml:4620]: Also renamed it `DomainTagAddress` for the time being

**Comment by h2xd** [comment]
Big supporter of this, as this gives not only more insights to the Frontend Engineers but also later customer that consume potentially the API. This would allow the customers to react to errors and build more recoverable services.

> it's great that you come up with the proposal. Do you have any idea how we can better review that? The view in GH is not optimal to comment :-)

Agree with @sfrehse here it's really hard to get into.

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 5]
```diff
@@ -0,0 +1,5677 @@
...
+            application/json:
+              schema:
+                $ref: '#/components/schemas/TagSchemaList'
+          description: Success
+        "401":
+          $ref: '#/components/responses/Unauthorized'
+      summary: Get data schemas
+      tags:
+        - Domain
+  /api/v1/management/domain/topics:
```
Comment (Florian-Limpoeck) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:3965]: is this only for publish 'topics' or also for subscribe 'topicFilters' ?

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 6]
```diff
@@ -0,0 +1,5677 @@
...
+            minimum: 500
+            default: 1000
+            maximum: 10000
+      responses:
+        "200":
+          content:
+            application/json:
+              example:
+                items:
+                  - client1/topic1
```
Comment (Florian-Limpoeck) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:4029]: i dont understand the examples. wouldnt it be something like:
{
  "clientId" : "my-fancy-client",
  "topic": "a/b/c"
}

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 7]
```diff
@@ -0,0 +1,5677 @@
...
+      summary: Get data schemas
+      tags:
+        - Domain
+
+  /api/v1/management/client/topic-samples:
+    get:
+      description: Get a sample of MQTT topics published by clients connected to the Edge broker over a period of time
+      operationId: get-client-topics
+      parameters:
+        - description: The time limit for the observation.
```
Comment (Florian-Limpoeck) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:4014]: in seconds or milliseconds?

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml (hunk 2 above)
Comment (Florian-Limpoeck) [reply ext/hivemq-edge-openapi-2024.8-proposal.yaml:4009]: why is this path under the root '/client' but topics are under the root '/domain' ?

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 8]
```diff
@@ -0,0 +1,5677 @@
...
+          in: query
+          name: queryTime
+          required: true
+          schema:
+            type: integer
+            minimum: 500
+            default: 1000
```
Comment (Florian-Limpoeck) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:4021]: in millis, the default seems a bit short, there can be nothing. in seconds its far to long, we cant cache topic data for so long

File: ext/hivemq-edge-openapi-2024.8-proposal.yaml [hunk 9]
```diff
@@ -0,0 +1,5677 @@
...
+                type: array
+                items:
+                  type: string
+          description: Success
+        "401":
+          $ref: '#/components/responses/Unauthorized'
+      summary: Get all topic tags
+      tags:
+        - Domain
+  /api/v1/management/domain/topics/schema:
```
Comment (Florian-Limpoeck) [inline ext/hivemq-edge-openapi-2024.8-proposal.yaml:3983]: what schema for what topic? Where does the schema come from? does it apply to all topics?
